pkg/scheduler/algorithm/priorities/scarce_resource_bin_packing.go

```diff
@@ -0,0 +1,126 @@
+*/
+
+package priorities
+
+import (
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
+)
+
+// ScarceResourceBinPacking contains information to calculate bin packing priority.
+type ScarceResourceBinPacking struct {
+	scarceResource string
+}
+
+// NewScarceResourceBinPacking creates a ScarceResourceBinPackingPriorityMap.
+func NewScarceResourceBinPacking(scarceResource string) (algorithm.PriorityMapFunction, algorithm.PriorityReduceFunction) {
+	scarceResourceBinPackingPrioritizer := &ScarceResourceBinPacking{
+		scarceResource: scarceResource,
+	}
+	return scarceResourceBinPackingPrioritizer.ScarceResourceBinPackingPriorityMap, nil
+}
+
+// ScarceResourceBinPackingPriorityDefault creates a requestedToCapacity based
+// ResourceAllocationPriority using default resource scoring function shape.
+// The default function assigns 1.0 to resource when all capacity is available
+// and 0.0 when requested amount is equal to capacity.
+func ScarceResourceBinPackingPriorityDefault() *ScarceResourceBinPacking {
+	defaultscarceResourceBinPackingPrioritizer := &ScarceResourceBinPacking{
+		scarceResource: "",
+	}
+	return defaultscarceResourceBinPackingPrioritizer
+}
+
+// ScarceResourceBinPackingPriorityMap is a priority function that favors nodes that have higher utlization of scare resource.
+// It will detect whether the requested scarce resource is present on a node, and then calculate a score ranging from 0 to 10
+// based total utlization (best fit)
+// - If none of the scare resource are requested, this node will be given the lowest priority.
+// - If the scarce resource is requested, the larger the resource utlization ratio, the higher the node's priority.
+func (s *ScarceResourceBinPacking) ScarceResourceBinPackingPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error) {
+	var score int
+	node := nodeInfo.Node()
+	if len(s.scarceResource) == 0 {
+		return schedulerapi.HostPriority{}, fmt.Errorf("scarceResource not defined")
+	}
+	if node == nil {
+		return schedulerapi.HostPriority{}, fmt.Errorf("node not found")
+	}
+	if !podRequestsResource(*pod, s.scarceResource) {
+		score = 0
+	} else {
+		score = calculateScareResourceScore(nodeInfo, pod.Spec.Containers, s.scarceResource)
+	}
+
+	return schedulerapi.HostPriority{
+		Host:  node.Name,
+		Score: score,
+	}, nil
+}
+
+// calculateScareResourceScore returns total utlization of the scare resource on the node
+func calculateScareResourceScore(nodeInfo *schedulercache.NodeInfo, containers []v1.Container, resource string) int {
+	reqResource := 0
+	usedResource := 0
+	for _, container := range containers {
+		if qunatity, ok := container.Resources.Requests[v1.ResourceName(resource)]; ok {
+			reqResource += int(qunatity.Value())
+		}
+	}
+	for _, pod := range nodeInfo.Pods() {
```

Comment:
Can we reuse `ResourceAllocationPriority` ?

Reply:
The ResourceAllocationPriority is centered around Memory and CPU only. It does not consider the extended resources. Thus the resource bin packing function. When the resource bin packing function supports multiple resources we could use this priority function.

Reply:
i briefly looked at the code and the algorithm is exactly similar to what is used in ResourceAllocationPriority except that it only accounts for cpu,mem and volumes. I bet we can refactor it to work with extended resources as well. That will reduce a lot of code duplication

Reply:
I still feel the need for a dedicated bin packing, the resource allocation is a wrapper function for least requested and most requested. Resource Bin Packing will eventually allow users to choose on which resources to bin pack. Modifying the resource allocation to consider all the extended resoures on the cluster does not sound a good idea to me.

Reply:
I am not very picky about doing it in this pr but I think they should merge eventually

Reply:
👍